grafana-plugin/e2e-tests/fixtures.ts

```diff
@@ -30,24 +34,47 @@ type Fixtures = {
   adminRolePage: AdminRolePage;
 };
 
-export * from '@playwright/test';
-export const test = base.extend<Fixtures>({
-  viewerRolePage: async ({ browser }, use) => {
-    const context = await browser.newContext({ storageState: VIEWER_USER_STORAGE_STATE });
-    const page = new ViewerRolePage(await context.newPage());
-    await use(page);
-    await context.close();
-  },
-  editorRolePage: async ({ browser }, use) => {
-    const context = await browser.newContext({ storageState: EDITOR_USER_STORAGE_STATE });
-    const page = new EditorRolePage(await context.newPage());
-    await use(page);
-    await context.close();
-  },
-  adminRolePage: async ({ browser }, use) => {
-    const context = await browser.newContext({ storageState: ADMIN_USER_STORAGE_STATE });
-    const page = new AdminRolePage(await context.newPage());
+/**
+ * NOTE: currently videos are not generated automatically because of how we generate a browserContext within our
+ * auth fixtures (which is how Playwright suggested setting up multi-role authnz tests..). There's a GitHub
+ * Issue here that tracks this issue https://github.com/microsoft/playwright/issues/24542
+ *
+ * Here's a temporary workaround on this, which is what this function does
+ * https://github.com/microsoft/playwright/issues/14813#issuecomment-1582499142
+ */
+const _recordTestVideo = async (
+  browser: Browser,
+  use: (r: BaseRolePage) => Promise<void>,
+  testInfo: TestInfo,
+  storageStateLocation: string,
+  RolePage: BaseRolePageType
+) => {
+  const videoDir = path.join(testInfo.outputPath(), 'videos');
+
+  const context = await browser.newContext({ storageState: storageStateLocation, recordVideo: { dir: videoDir } });
+  const page = new RolePage(await context.newPage());
+
+  try {
     await use(page);
+  } finally {
     await context.close();
-  },
+    const videoFiles = fs.readdirSync(videoDir);
+
+    if (videoFiles.length > 0) {
+      for (let i = videoFiles.length; i > 0; i--) {
+        let videoFile = path.join(videoDir, videoFiles[i - 1]);
+        await testInfo.attach('video', { path: videoFile });
+      }
+    }
+  }
+};
+
+export * from '@playwright/test';
+export const test = base.extend<Fixtures>({
+  viewerRolePage: ({ browser }, use, testInfo) =>
+    _recordTestVideo(browser, use, testInfo, VIEWER_USER_STORAGE_STATE, ViewerRolePage),
+  editorRolePage: async ({ browser }, use, testInfo) =>
+    _recordTestVideo(browser, use, testInfo, EDITOR_USER_STORAGE_STATE, EditorRolePage),
+  adminRolePage: async ({ browser }, use, testInfo) =>
+    _recordTestVideo(browser, use, testInfo, ADMIN_USER_STORAGE_STATE, AdminRolePage),
```

Comment:
this fixes the issue where once we switched to fixture based `page` objects, `playwright` stopped adding videos to `playwright` test artifacts. These are super useful for debugging broken/flaky tests. See [this github issue](https://github.com/microsoft/playwright/issues/14813) for more context.